### PR TITLE
RBL code cleanup

### LINKED ISSUE #7346: RBL/Debugging code cleanup
- Look at how the debug adapter tracker running the variables view interacts with the KernelDebugAdapter
- Split KernelDebugAdapter into multiple classes that only have one job
  - https://github.com/microsoft/vscode-jupyter/pull/7416
- Move the DebugAdapterDescriptorFactory into its own class (depends on variables stuff, and not setting global context keys for debug)
- We send disconnect multiple times which triggers errors from the kernel

### PR DESCRIPTION
**Pull request by roblourens**
Fix #7346

This moves most of the RBL and cell debug specific code out of KernelDebugAdapter into controller classes. There is some left that will be a separate change for the variables changes we discussed in #7346. 

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] ~~Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).~~
-   [x] Appropriate comments and documentation strings in the code.
-   [x] Has sufficient logging.
-   [x] Has telemetry for enhancements.
-   [x] Unit tests & system/integration tests are added/updated.
-   [x] ~~[Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.~~
-   [x] ~~[`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).~~

### CODE REVIEW

File: src/client/debugger/jupyter/kernelDebugAdapter.ts [hunk 1]
```diff
@@ -110,30 +104,16 @@ export class KernelDebugAdapter implements DebugAdapter, IKernelDebugAdapter, ID
...
-
                 if (anyMsg.header.msg_type === 'debug_event') {
-                    if (anyMsg.content.event === 'stopped') {
-                        this.threadId = anyMsg.content.body.threadId;
-                        if (await this.handleStoppedEvent()) {
-                            this.trace('intercepted', JSON.stringify(anyMsg.content));
-                            return;
-                        }
+                    this.trace('event', JSON.stringify(msg));
+                    if (!(await this.delegate?.willSendEvent(anyMsg))) {
```
Comment (IanMatthewHuff) [inline src/client/debugger/jupyter/kernelDebugAdapter.ts:114]: Nice feels much better this way.

**Review by IanMatthewHuff — APPROVED** [review]

**Review by DavidKutu — APPROVED** [review]

File: src/client/debugger/jupyter/debugControllers.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            if (await this.handleStoppedEvent(this.lastPausedThreadId!)) {
+                this.trace('intercepted', JSON.stringify(anyMsg.content));
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public async willSendRequest(request: DebugProtocol.Request): Promise<void> {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugControllers.ts:85]: Not sure i like this `willSendrequest`.
`will` implies there's a question, and I'd expect the response to be `true/false` or similar.
But I'm not seeing a return value, hence this looks more like a command to do somethign as opposed to a method used to inquire somethign.

Comment (DonJayamanne) [reply src/client/debugger/jupyter/debugControllers.ts:85]: Never mind, read the comments in interface.
Do we need an interface

Comment (roblourens) [reply src/client/debugger/jupyter/debugControllers.ts:85]: Should I call them `onWill...`?

Comment (roblourens) [reply src/client/debugger/jupyter/debugControllers.ts:85]: And the fact that one returns a boolean while the other doesn't may or may not be bad, I can't decide

**Pull request merged**